File: packages/core/src/platform/logger.ts
/**
 * Minimal structured logger. Emits JSON lines in production (parseable by any
 * log pipeline) and pretty lines in dev. The interface is what code depends on,
 * so swapping in pino/winston/OTel later is a one-file change.
 */

export type LogFields = Record<string, unknown>;

export interface Logger {
  debug(msg: string, fields?: LogFields): void;
  info(msg: string, fields?: LogFields): void;
  warn(msg: string, fields?: LogFields): void;
  error(msg: string, fields?: LogFields): void;
  /** Returns a logger that merges `fields` into every line. */
  child(fields: LogFields): Logger;
}

type Level = "debug" | "info" | "warn" | "error";
const ORDER: Record<Level, number> = { debug: 0, info: 1, warn: 2, error: 3 };

export interface LoggerOptions {
  level?: Level;
  pretty?: boolean;
  base?: LogFields;
}

export function createLogger(opts: LoggerOptions = {}): Logger {
  const level = opts.level ?? (process.env.NODE_ENV === "production" ? "info" : "debug");
  const pretty = opts.pretty ?? process.env.NODE_ENV !== "production";
  const base = opts.base ?? {};

  function emit(lvl: Level, msg: string, fields?: LogFields): void {
    if (ORDER[lvl] < ORDER[level]) return;
    const record = { level: lvl, time: new Date().toISOString(), msg, ...base, ...fields };
    if (pretty) {
      const extra = { ...base, ...fields };
      const tail = Object.keys(extra).length ? " " + JSON.stringify(extra) : "";
      // eslint-disable-next-line no-console
      console[lvl === "debug" ? "log" : lvl](`[${lvl}] ${msg}${tail}`);
    } else {
      // eslint-disable-next-line no-console
      console[lvl === "debug" ? "log" : lvl](JSON.stringify(record));
    }
  }

  return {
    debug: (m, f) => emit("debug", m, f),
    info: (m, f) => emit("info", m, f),
    warn: (m, f) => emit("warn", m, f),
    error: (m, f) => emit("error", m, f),
    child: (fields) => createLogger({ level, pretty, base: { ...base, ...fields } }),
  };
}

File: apps/api/internal/platform/logger.go
package platform

import (
	"log/slog"
	"os"
)

// Logger is a minimal structured logger interface; what code depends on, so the
// backing implementation (slog here) can be swapped. Mirrors
// packages/core/src/platform/logger.ts.
type Logger interface {
	Debug(msg string, args ...any)
	Info(msg string, args ...any)
	Warn(msg string, args ...any)
	Error(msg string, args ...any)
	// Child returns a logger that merges args into every line.
	Child(args ...any) Logger
}

type slogLogger struct{ l *slog.Logger }

func (s slogLogger) Debug(msg string, args ...any) { s.l.Debug(msg, args...) }
func (s slogLogger) Info(msg string, args ...any)  { s.l.Info(msg, args...) }
func (s slogLogger) Warn(msg string, args ...any)  { s.l.Warn(msg, args...) }
func (s slogLogger) Error(msg string, args ...any) { s.l.Error(msg, args...) }
func (s slogLogger) Child(args ...any) Logger      { return slogLogger{l: s.l.With(args...)} }

// NewLogger builds a logger. Pretty (text) output in dev, JSON in production,
// matching the TS logger's environment behavior.
func NewLogger(pretty bool, base ...any) Logger {
	level := slog.LevelInfo
	if pretty {
		level = slog.LevelDebug
	}
	opts := &slog.HandlerOptions{Level: level}
	var h slog.Handler
	if pretty {
		h = slog.NewTextHandler(os.Stderr, opts)
	} else {
		h = slog.NewJSONHandler(os.Stderr, opts)
	}
	l := slog.New(h)
	if len(base) > 0 {
		l = l.With(base...)
	}
	return slogLogger{l: l}
}
